### PR TITLE
ReplayGain support + audio normalization (web player)

### LINKED ISSUE #233: Volume normalization (in web player)
Volume is usually a big problem in libraries. When volume is normalized tools like MP3Gain and I believe also replaygain add information to the ID3 tag without touching the data itself. So just analyzing the volume and adding metadata. I think we should have a setting to enable support for that at least in the webplayer. I don't know exactly what tags are used for that but I suggest that we research that and use that information.

I have some ideas on the implementation (and I have no idea if it's a common implementation like this, just came to my mind).

Lets say we have three songs with different volume levels:
- Song A: 25
- Song B: 50
- Song C: 75

Assuming the normal level should be 50, you'd expect Song A to have a metadata tag saying +25 and song C having a tag saying -25.

Now I suggest we introduce two volume levels in the Web Player. One Hidden and one Shown. The Shown one is simply for the user to control the output volume from 0-100%.

The hidden one can be used for the volume correction according to the metadata tag. So when our hidden volume is 50, we can use the tag to add or remove 25 from the baseline volume depending on the metadata tag but it won't affect the volume setting of the users 0-100% which is just an overall volume control.

_Regarding streaming capabilities to other devices and when other apps are used over the Airsonic API, I have no solution to fix this yet without actually changing the data (during transcoding)._

### LINKED ISSUE #1850: Expose replaygain tags
Hey,

I know that replaygain support in general is an already requested feature.
But it would be great if replaygain tags could be exposed to just start somewhere.

The dev of Symfonium seems to be [open about supporting](https://support.symfonium.app/t/replaygain-support/265/3) it if the tags are exposed (as far as I know the local part of the app already supports it anyway).

That would at least give us one setup constellation consisting of two popular projects, it maybe also could kick-start the whole thing.

Thanks for your work!

### PR DESCRIPTION
**Pull request by kgarner7**
resolve #1850, resolve #233, and part of #1036

This PR seeks to add extraction of ReplayGain tags, the ability to get this data via the native API, and audio normalization in the web player.

The web player is unchanged, I have just elected to use the Web Audio API to wrap around the `<audio>` element. I'm not sure if the gainNode with values > 1 is problematic or not yet (maybe they need to be normalized to 1?).

I'd be happy to have some feedback/additional testing (or if someone has a different player, now might be a good time to switch).

### CODE REVIEW

**Comment by deluan** [comment]
Nice! I'll take a look and give you feedback on this as soon as I can. Thanks!

**Comment by deluan** [comment]
Hey @kgarner7 , does this take into consideration transcoding? I think we need to make `ffmpeg` add the RG tags in the output, so other players would receive the tags.

**Comment by kgarner7** [comment]
I'm not entirely sure how transcoding would play a role. Does that change the replaygain properties of the track (e.g., if I started with flac and converted it to ogg, would those values change)? The tags come from the original file (and are stored in the db).

**Comment by deluan** [comment]
I may be wrong, but I think we just have to reuse the tags from the original file in the transcoded one. I think `ffmpeg` handles this. We have to dig a bit more. We can always have a following PR to handle that.

**Comment by kgarner7** [comment]
Ah, if the concern is whether ffmpeg will expose those same tags to the new file, I believe the answer is yes (when I transcoded my library to Play:Sub, which does replaygain, I could see gain info).

**Comment by rsammelson** [comment]
I play my music transcoded from FLAC to Opus, and the gain was working correctly in the web player.

**Comment by deluan** [comment]
@rsammelson My questioning is about other clients, the Subsonic clients. Will they benefit from this? Base on @kgarner7 , seems that this works as well

**Comment by kgarner7** [comment]
Ahhh,I understand now. I don't think so. The reason why Play:Sub worked was because it read the tags from file. It isn't exposed on subsonic routes (I guess I could modify childFromMediaFile?), but I'm not sure that is used...

**Comment by deluan** [comment]
> Ahhh,I understand now. I don't think so. The reason why Play:Sub worked was because it read the tags from file. It isn't exposed on subsonic routes (I guess I could modify childFromMediaFile?), but I'm not sure that is used...

Yes, that's how Subsonic clients handle ReplayGain. We just need to be sure that whatever RG tags are in the files, they remain intact after transcoding. My understanding is that is what currently happens (based on your test with play:Sub), right? To be sure, you could also download a track with transcoding ON and check if `ffmpeg -i downloaded_track.mp3` has the same tags as the original.

File: db/migration/20221113142454_add_replaygain_metadata.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	_, err := tx.Exec(`
+alter table media_file add 
+	album_gain real;
+alter table media_file add 
+	album_peak real;
+alter table media_file add 
+	track_gain real;
+alter table media_file add 
+	track_peak real;
+`)
```
Comment (deluan) [inline db/migration/20221113142454_add_replaygain_metadata.go:23]: Can you please add a `rg_` prefix to these columns?

File: model/mediafile.go [hunk 2]
```diff
@@ -9,50 +9,55 @@ type MediaFile struct {
...
+	MbzReleaseTrackID    string  `structs:"mbz_release_track_id" json:"mbzReleaseTrackId,omitempty" orm:"column(mbz_release_track_id)"`
+	MbzAlbumID           string  `structs:"mbz_album_id" json:"mbzAlbumId,omitempty"         orm:"column(mbz_album_id)"`
+	MbzArtistID          string  `structs:"mbz_artist_id" json:"mbzArtistId,omitempty"        orm:"column(mbz_artist_id)"`
+	MbzAlbumArtistID     string  `structs:"mbz_album_artist_id" json:"mbzAlbumArtistId,omitempty"   orm:"column(mbz_album_artist_id)"`
+	MbzAlbumType         string  `structs:"mbz_album_type" json:"mbzAlbumType,omitempty"`
+	MbzAlbumComment      string  `structs:"mbz_album_comment" json:"mbzAlbumComment,omitempty"`
+	AlbumGain            float64 `structs:"album_gain" json:"albumGain"`
+	AlbumPeak            float64 `structs:"album_peak" json:"albumPeak"`
+	TrackGain            float64 `structs:"track_gain" json:"trackGain"`
+	TrackPeak            float64 `structs:"track_peak" json:"trackPeak"`
```
Comment (deluan) [inline model/mediafile.go:57]: Same here, please add a `RG` prefix (ex: `RGAlbumGain`)

File: scanner/metadata/metadata.go [hunk 3]
```diff
@@ -120,6 +120,39 @@ func (t Tags) Size() int64                 { return t.fileInfo.Size() }
 func (t Tags) FilePath() string            { return t.filePath }
 func (t Tags) Suffix() string              { return strings.ToLower(strings.TrimPrefix(path.Ext(t.filePath), ".")) }
 
+// Replaygain Properties
+func (t Tags) AlbumGain() float64 { return t.getGainValue("replaygain_album_gain") }
+func (t Tags) AlbumPeak() float64 { return t.getPeakValue("replaygain_album_peak") }
+func (t Tags) TrackGain() float64 { return t.getGainValue("replaygain_track_gain") }
+func (t Tags) TrackPeak() float64 { return t.getPeakValue("replaygain_track_peak") }
+
```
Comment (deluan) [inline scanner/metadata/metadata.go:128]: Same here (prefix RG)

**Comment by kgarner7** [comment]
I downloaded a transcoded track, and I can confirm that the replaygain attributes are still there (I also manually transcoded with ffmpeg for same results).

File: ui/src/common/SongInfo.js [hunk 4]
```diff
@@ -17,12 +17,24 @@ import inflection from 'inflection'
...
 import { makeStyles } from '@material-ui/core/styles'
+import config from '../config'
 
-const useStyles = makeStyles({
-  tableCell: {
-    width: '17.5%',
-  },
-})
+const useStyles = config.enableReplayGain
+  ? makeStyles({
```
Comment (deluan) [inline ui/src/common/SongInfo.js:23]: There's a better (more MUI) way to achieve this: https://github.com/navidrome/navidrome/blob/31c598de075fc7af96be93070056a6666da1b5c3/ui/src/common/SongDatagrid.js#L48

**Review by deluan — CHANGES_REQUESTED** [review]
This looks great! Just need to test it locally and see how the UX feels.

**Comment by kgarner7** [comment]
Additional feature: ability to toggle preamp gain

File: ui/src/i18n/en.json [hunk 5]
```diff
@@ -363,7 +363,14 @@
...
                 "lastfmScrobbling": "Scrobble to Last.fm",
-                "listenBrainzScrobbling": "Scrobble to ListenBrainz"
+                "listenBrainzScrobbling": "Scrobble to ListenBrainz",
+                "replaygain": "ReplayGain Mode",
+                "preAmp": "ReplayGain PreAmp (dB)",
+                "gain": {
+                    "none": "Disabled",
+                    "album": "Normalize albums",
+                    "track": "Normalize tracks"
+                }
```
Comment (kgarner7) [inline ui/src/i18n/en.json:373]: Suggestions for better descriptions here are appreciated

Comment (metalheim) [reply ui/src/i18n/en.json:373]: I'd prefer "Use Album Gain" and "Use Track Gain". 
Thats f.e. how the DSub Android Client states it

Comment (deluan) [reply ui/src/i18n/en.json:373]: Yep, makes sense @metalheim

**Review by deluan — APPROVED** [review]

**Pull request merged**